### ISSUE TITLE
Decide how public-facing instance traits should be

### ISSUE THREAD

**Issue by travisbrown** [typelevel/cats#3008]
Right now we have `AllInstances` and `AllSyntax` traits that are in the public API and are sometimes used outside of Cats itself—for example in Circe we have a `CirceSuite` helper class that extends these traits so that we don't have to import `cats.implicits._` in every test file.

The problem with this kind of usage is that `AllInstances` and `AllSyntax` don't actually include all of anything, ever since we started stacking on `BinCompatN` layers to maintain binary compatibility on Scala 2.11. For example, want a `Defer` for `Function0`? You won't get it from `AllInstances`—you also need `AllInstancesBinCompat0`.

This means anyone using `AllInstances` or `AllSyntax` has to be careful to add the latest `AllSyntaxBinCompat37` or whatever traits included in each new Cats release if they actually want everything.

This is something we haven't managed to stay on track on even within Cats itself—e.g. see #3007, which is due to the fact that `cats.implicits` is lagging two `AllSyntaxBinCompat`s behind, or `SyntaxSuite`, which is four `AllInstanceBinCompat`s out of date (but happens to still be working), etc.

To make things even more confusing, there's a `AllSyntaxBinCompat` (note the missing number) abstract class that bundles up all the syntax bin-compat traits, but no equivalent `AllInstancesBinCompat`.

I've also recently proposed (and merged) making the bin-compat traits package-private (#3003), which on second thoughts I think we'll have to undo, at least until 2.1.

So there's a lot of miscellaneous little brokenness and inconsistency and mildly bad UX here, some of which I'm about to open a PR to address, but there's also a bigger question of whether we actually intend the `AllInstances` and `AllSyntax` traits to be used outside of the implementation of Cats itself.

**Comment by Daenyth** [comment]
Just as a data point, I've used those in third-party code so that I could make one combined import for cats implicits + my library implicits + another library implicits, to help reduce brain effort on imports

**Comment by djspiewak** [comment]
As another random data point… I originally structured djspiewak/shims to hide the "All" traits, so the only way to actually get at anything was via the `shims._` import. The problem with this is it made it impossible for any third parties to participate in the implicit prioritization, which created unavoidable conflicts. Given Cats' nature, I would suspect similar problems if we privatized the traits.

**Comment by joroKr21** [comment]
But this is only relevant for syntax now, isn't it?

**Comment by djspiewak** [comment]
> But this is only relevant for syntax now, isn't it?

Yes, which significantly decreases (though does not entirely remove) the relevancy of interaction with prioritization.
